CLN: use integer parsing functions from stdlib

Alvaro-Kothe opened the pull request:
This was a suggestion from @WillAyd in https://github.com/pandas-dev/pandas/pull/62623#discussion_r2421930388

This simplifies the current parsing logic by utilizing existing functions to parse integers from stdlib (`strtoll` and `strtoull`) instead of using our own.

~One disadvantage of this approach is that in case there is a thousand separator (`tsep`) we allocate memory on the heap and process the existing string to remove this separator.~

Alvaro-Kothe committed:
refactor(parser): use integer parsing functions from stdlib

WillAyd reviewed (changes requested):
I think this is much more difficult and less performant than it needs to be because we have a suboptimal structure for the parser.

Instead of storing a `char **words` member, is there a way that we can define a struct like:

```c
struct PandasParsedString {
    const char *str;
    size_t len;
    int64_t[2] is_separator;
}
```

and then make the member in the parser a `PandasParsedString *words` array. As you parse each word you can assign these struct members once, instead of having to query them after the fact.

I'm assuming the `is_separator` member is a bit-packed struct signifying whether a particular character is a separator or not (see other comment about a 128 character limit). We could implement this a few other ways, but I figure this is the most memory efficient.

With that information up front you can avoid having to do a character search for separators here way after the fact; if there are no separators you can avoid looping altogether, but if there are you can move a region of characters in batches to the local buffer rather than having to go char by char

Alvaro-Kothe committed:
perf: use a local buffer to store the processed string

Alvaro-Kothe committed:
fix: use macro to fix MSVC build error

Alvaro-Kothe committed:
fix: use `bool`

Alvaro-Kothe committed:
refactor: don't pass PROCESSED_WORD_CAPACITY as a separate argument

Alvaro-Kothe committed:
perf: write in chunks

Alvaro-Kothe committed:
hack: try bigger buffer size for arm error

Alvaro-Kothe committed:
fix: solution without manipulating the string

Alvaro-Kothe committed:
some cleanup

Alvaro-Kothe committed:
fix: use ptrdiff_t to fix MSVC build error

Alvaro-Kothe committed:
add other exponent cases for completion

Alvaro-Kothe committed:
fix: use builtin overflow check verification

Alvaro-Kothe committed:
fix: change std to c2x

Alvaro-Kothe committed:
Revert previous commits

Revert "fix: change std to c2x"

This reverts commit 9046eccb8e49c945ac9f6184fa8a9a54944372e7.

Revert "fix: use builtin overflow check verification"

This reverts commit d85aaf02bd85f12459ca7dd0039f4a9a1ee468b5.

Alvaro-Kothe committed:
refactor: move overflow check to header

Alvaro-Kothe committed:
refactor: use overflow check from numpy

Alvaro-Kothe committed:
fix: handle negative check

Alvaro-Kothe committed:
fix: add test for thousand separator with negative number

Alvaro-Kothe committed:
move to portable

Alvaro-Kothe committed:
perf: use builtin unsigned long overflow check

Alvaro-Kothe committed:
refactor: combine builting and gnuc branches

Alvaro-Kothe committed:
don't assign null

Alvaro-Kothe committed:
fix: perform bound check

Alvaro-Kothe committed:
fix: assign error if doesn't have a digit after tsep

It stills permits ending in tsep

Alvaro-Kothe committed:
fix: go back to buffer solution

Alvaro-Kothe committed:
refactor: undo refactor in np_datetime.c

Alvaro-Kothe committed:
fix: fix undefined behavior

Alvaro-Kothe committed:
fix: fix leftover undefined behaviour

Alvaro-Kothe committed:
rewrite `copy_string_without_char`

- Returns status code
- Simplify loop for a more common sliding window
- use string length

Alvaro-Kothe committed:
fix: change solution to safe guard against end_ptr

Alvaro-Kothe committed:
test: add some edge cases tests with thousand separator

WillAyd reviewed (changes requested):
nice work - headed in the right direction

Alvaro-Kothe committed:
Update pandas/_libs/src/parser/tokenizer.c

Co-authored-by: William Ayd <william.ayd@icloud.com>

Alvaro-Kothe committed:
fix: error to -1

Alvaro-Kothe committed:
fix: leftover status check

Alvaro-Kothe committed:
fix: remove duplicate nbytes declaration

WillAyd reviewed (commented):
Will have to do another pass tomorrow but generally this looks great - thanks for sticking with all of the feedback.

Have you benchmarked in its current state?

Alvaro-Kothe committed:
fix: use memchr to find if need to process the word

Alvaro-Kothe committed:
chore: add comment explaining why 128 bytes for capacity

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1837,197 @@ int uint64_conflict(uint_state *self) {
-            ((number == pre_max) && (d - '0' <= dig_pre_max))) {
-          number = number * 10 + (d - '0');
-          d = *++p;
+  errno = 0;
```
I think you can remove this now, right? Generally I'd advise against clearing `errno` without good reason. This currently is equivalent to doing a `try ... except Exception` in Python and blindly continuing along, so it should be avoided

Alvaro-Kothe committed:
rename bytes_read to bytes_written

Alvaro-Kothe committed:
chore: move errno assignment

Alvaro-Kothe committed:
fix: fix error logic by comparing pointers

Alvaro-Kothe committed:
fix: use pointers

Alvaro-Kothe committed:
fix: keep track on how many bytes to read

Alvaro-Kothe committed:
chore: cast to size_t

Alvaro-Kothe committed:
fix: cast pointer to fix Wc++-compat warning

Alvaro-Kothe commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,191 @@ int uint64_conflict(uint_state *self) {
+  size_t bytes_written = 0;
+
+  while (left < end_ptr) {
+    right = memchr(left, char_to_remove, end_ptr - left);
```
I updated the code to keep track of how many bytes it read. I also compiled with `-Wall` and made some casts to remove some warnings of `-Wc++-compat` due to `memchr` and `-Wsign-conversion` due to implicit casting of `ptrdiff_t`.

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
+    // We are also casting directly to size_t because
+    // `left` never goes beyond `right` or `end_ptr`.
+    size_t chunk_size =
+        right ? (size_t)(right - left) : (size_t)(end_ptr - left);
```
```suggestion
        right ? right - left : end_ptr - left;
```

No point in casting these - the assignment will do it

Alvaro-Kothe committed:
fix: remove casts for -Weverything

Alvaro-Kothe committed:
fix: move remaining_bytes_to_read to the start of loop

Alvaro-Kothe commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
+    bytes_written += chunk_size;
+
+    // Advance past the removed character if we found it.
+    left = right ? right + 1 : end_ptr;
```
I moved the remaining_bytes_to_read to the start of the loop. For me, the branching is quite clean, using only 2 ternary operators.

To consolidate it even further, I think it would need an explicit `if-else` and a temporary variable to indicate where left should point to.

All these ternaries are mainly because I am avoiding `left` pointing to `end_ptr + 1`.

WillAyd replied:
> All these ternaries are mainly because I am avoiding `left` pointing to `end_ptr + 1`.

There's nothing wrong with this pointing to that location as long as you don't try to subsequently read or write it

Alvaro-Kothe replied:
Now I reassign right to end_ptr if it's NULL.

Alvaro-Kothe committed:
fix: consolidate it even further

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
+                                    const char *str, size_t str_len,
+                                    char char_to_remove) {
+  const char *left = str;
+  const char *right;
```
Since `right` is only used inside of the loop you can move this therein (to the initial definition)

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
+      right = end_ptr;
+    }
+
+    size_t chunk_size = right - left;
```
```suggestion
    const size_t chunk_size = right - left;
```

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
+
+    size_t chunk_size = right - left;
+
+    // check if we have enough space, including the null terminator.
```
I think the comments here and beyond are all superfluous, i.e. they add no value on top of what the code itself is showing

Alvaro-Kothe committed:
fix: move right definition

Alvaro-Kothe committed:
chore: remove superfluous comments

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
-  // Skip trailing spaces.
-  while (isspace_ascii(*p)) {
-    ++p;
+  if (errno == ERANGE) {
```
This is lost in a refactor right? I don't think errno will be set by anything preceding this?

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
+  char *endptr = NULL;
+  // strtoll sets errno if it finds an overflow.
+  // It's value is reset to don't pollute the verification below.
+  errno = 0;
```
As a matter of convention need to check `errno` and "raise" if its set before clearing it

Alvaro-Kothe replied:
By "raise" you mean printing to stderr?

Alvaro-Kothe replied:
I am now resetting `errno` after handling the overflow error

Alvaro-Kothe committed:
fix: add const qualifier

Alvaro-Kothe commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
+  const char *end_ptr = str + str_len;
+  size_t bytes_written = 0;
+
+  while (left < end_ptr) {
```
It's not necessary, because it will be a no-op. It reads 0 bytes from `remaining_bytes_to_read = end_ptr - left;` and copies 0 bytes with `memcpy`.

Alvaro-Kothe committed:
fix: remove unnecessary NULL and null-byte checks

Alvaro-Kothe committed:
fix: remove unnecessary errno verification

Alvaro-Kothe committed:
chore: remove NULL assignment

Alvaro-Kothe commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,196 @@ int uint64_conflict(uint_state *self) {
-  // Skip leading spaces.
-  while (isspace_ascii(*p)) {
-    ++p;
+  if (!p_item || *p_item == '\0') {
```
Indeed, not necessary. The null byte is already handled below.

Alvaro-Kothe committed:
fix: don't recompute strlen

Alvaro-Kothe committed:
chore: add some comments back to simplify diff

Alvaro-Kothe committed:
fix: reset errno after handling it

Alvaro-Kothe committed:
fix: put back const char p

Alvaro-Kothe commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1834,201 +1839,177 @@ int uint64_conflict(uint_state *self) {
   // Skip leading spaces.
-  while (isspace_ascii(*p)) {
-    ++p;
+  while (isspace_ascii(*p_item)) {
```
I've made several changes that complicate a lot the diff. I'll put back the `const char *p` assignment, and also remove the need for some auxiliary functions created.

WillAyd replied:
Great - thanks!

Alvaro-Kothe committed:
fix: improve diff for sign handling

Alvaro-Kothe committed:
fix: improve diff for trailing whitespace

Alvaro-Kothe committed:
chore: remove newline to simplify diff even more

Alvaro-Kothe committed:
chore: drop another superfluous comment

Alvaro-Kothe committed:
test: xfail python engine with consecutive thousand separators

Alvaro-Kothe committed:
fix: move errno handling to avoid polution and early return

Alvaro-Kothe committed:
chore: rename to number for diff

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1843,105 +1881,45 @@ int64_t str_to_int64(const char *p_item, int64_t int_min, int64_t int_max,
-        }
-      }
-    }
+  if (errno == ERANGE || number > int_max || number < int_min) {
```
```suggestion
  if (errno == ERANGE || number > int_max || number < int_min) {
```

Are you doing a follow up to get rid of the `int_max` and `int_min` arguments? Its strange to have this in the diff  given neither `number > INT64_MAX` nor `number < INT64_MIN` will ever be true, but I suppose within the larger context of there being an unnecessary argument for those I can see why you kept

WillAyd reviewed (approved):
lgtm. @mroeschke anything to add here?

Alvaro-Kothe committed:
chore: add comment explaining consecutive thousand separators in C engine

mroeschke commented:
Thanks @Alvaro-Kothe

Alvaro-Kothe commented:
Thanks for the patience @WillAyd. I already created the PR that changes the function signature in #62714.
